### PR TITLE
`PositiveDouble` prototype: representing a positive and finite numbers

### PR DESCRIPTION
**Pull request by ILIYANGERMANOV**
Based on #4 I've implemented a simple `PositiveDouble` prototype.

**TL;DR;**
- Unsafe creation via `PositiveDouble(5.0)` or `PositiveDouble.unsafe(-2.0)` which throws are exception
- Safe creation using `PositiveDouble.fromDouble(a)` which returns Option
- Overloads for +, -, *, /, %
- `PositiveDouble#mapSafe` extension for safely transforming the wrapped Double value
- `ArrowExactConstraintViolationError` with detailed error message for easier debugging or handling programatically

### CODE REVIEW

File: src/commonMain/kotlin/ArrowExactConstraintViolationError.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * )
+ *
+ * try {
+ *     // code that may throw a MyCustomConstraintError
+ * } catch (e: MyCustomConstraintError) {
+ *     println(e.message) // "Value 'some_value' does not meet the constraint 'MyCustomConstraint', which requires a custom constraint requirement."
+ * }
+ * ```
+ */
+open class ArrowExactConstraintViolationError(
```
Comment (ILIYANGERMANOV) [inline src/commonMain/kotlin/ArrowExactConstraintViolationError.kt:31]: Should we make this less verbose? It's a trade-off between information and verboseness.

File: src/commonMain/kotlin/number/PositiveDouble.kt [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+ * ### Example:
+ * ```
+ * try {
+ *     val invalidPositiveDouble = PositiveDouble(-1.0)
+ * } catch (e: ArrowExactPositiveDoubleConstraintError) {
+ *     println(e.message) // "Value '-1.0' does not meet the constraint 'PositiveDouble', which requires a positive and finite number."
+ * }
+ * ```
+ */
+class ArrowExactPositiveDoubleConstraintError(
```
Comment (ILIYANGERMANOV) [inline src/commonMain/kotlin/number/PositiveDouble.kt:214]: Should we make this less verbose? It's a trade-off between information and verboseness.

File: build.gradle.kts [hunk 3]
```diff
@@ -69,3 +69,7 @@ kotlin {
     }
   }
 }
+
+tasks.named<Test>("jvmTest") {
```
Comment (ILIYANGERMANOV) [inline build.gradle.kts:73]: Tests couldn't execute so I added this. Idk, if this or the invalidate & restart fixed the issues

**Comment by pacher** [comment]
Are you sure about operators returning `Option`? How would the usage look like for more complex formulas with more than one operation? `(a - b) * c - d / e` etc. It is a primitive, so I guess something like this should be easily doable without dozens of `bind` in between.
I was thinking more about something along the lines of
```kotlin
context(Raise<ArrowExactConstraintViolationError>)
operator fun plus(other: PositiveDouble): PositiveDouble
```
and dsl like `positive { (a - b) * c - d / e }` returning `Either<ArrowExactConstraintViolationError, PositiveDouble>`
Can't we do that?

Yes, I am suggesting to design it with context receivers from the start and keep experimental/ pre-release until context receivers are landed. I just don't see it any other way for primitives.

P.S. I apologize in advance, I could be totally off track here and misunderstand a lot since I did not even use arrow yet, just reading the new web site.

**Comment by ILIYANGERMANOV** [comment]
Side note: Don't forget to check #4 which our discussion issue:
**[Arrow Exact discussion](https://github.com/arrow-kt/arrow-exact/issues/4)**

Feel free to share your thoughts, ideas, and opinions - all are more than welcome!

**Comment by ILIYANGERMANOV** [comment]
> Are you sure about operators returning `Option`? How would the usage look like for more complex formulas with more than one operation? `(a - b) * c - d / e` etc. It is a primitive, so I guess something like this should be easily doable without dozens of `bind` in between. I was thinking more about something along the lines of
> 
> ```kotlin
> context(Raise<ArrowExactConstraintViolationError>)
> operator fun plus(other: PositiveDouble): PositiveDouble
> ```
> 
> and dsl like `positive { (a - b) * c - d / e }` returning `Either<ArrowExactConstraintViolationError, PositiveDouble>` Can't we do that?
> 
> Yes, I am suggesting to design it with context receivers from the start and keep experimental/ pre-release until context receivers are landed. I just don't see it any other way for primitives.
> 
> P.S. I apologize in advance, I could be totally off track here and misunderstand a lot since I did not even use arrow yet, just reading the new web site.

Good point @pacher! I like the `Raise` idea + Either DSL :+1: Let's discuss in #4. I hope that during the week more people will join and we'll get more POVs - which is normal I'm spamming people with code on Sunday (I usually rest on Sundays but got excited about the Arrow Exact library :smile: )

**Comment by pacher** [comment]
Also, it _feels_ right to do overloading like
```kotlin
operator fun plus(other: PositiveDouble) = if (...) throw ...

context(Raise<ArrowExactConstraintViolationError>)
operator fun plus(other: PositiveDouble) = if (...) raise ...
```
but it is actively [discouraged](https://github.com/Kotlin/KEEP/blob/master/proposals/context-receivers.md#overloading-by-the-presence-of-context) ... sigh

**Comment by pacher** [comment]
And the biggest design question is what to do when mixing with regular primitives. What should be the return type of `PositiveDouble - Double`? Should it be `PositiveDouble` and fail when the result is negative? Or should it be just `Double`? Both seem sound with valid use cases.

**Comment by ILIYANGERMANOV** [comment]
We decided to use a more generalized design in #4 which makes this PR obsolete.

**Pull request closed without merging**